sklearn/tree/_splitter.pyx

```diff
@@ -1,33 +1,49 @@
+"""Splitting algorithms in the construction of a tree.
+
+This module contains the main splitting algorithms for constructing a tree.
+Splitting is concerned with finding the optimal partition of the data into
+two groups. The impurity of the groups is minimized, and the impurity is measured
+by some criterion, which is typically the Gini impurity or the entropy. Criterion
+are implemented in the ``_criterion`` module.
+
+Splitting evaluates a subset of features (defined by `max_features` also
+known as mtry in the literature). The module supports two primary types
+of splitting strategies:
+
+- Best Split: A greedy approach to find the optimal split. This method
+  ensures that the best possible split is chosen by examining various
+  thresholds for each candidate feature.
+- Random Split: A stochastic approach that selects a split randomly
+  from a subset of the best splits. This method is faster but does
+  not guarantee the optimal split.
+"""
 # Authors: The scikit-learn developers
 # SPDX-License-Identifier: BSD-3-Clause
 
-from cython cimport final
-from libc.math cimport isnan
-from libc.stdlib cimport qsort
 from libc.string cimport memcpy
 
+from ._partitioner cimport DensePartitioner, SparsePartitioner, FEATURE_THRESHOLD
 from ._criterion cimport Criterion
-from ._utils cimport log
 from ._utils cimport rand_int
 from ._utils cimport rand_uniform
 from ._utils cimport RAND_R_MAX
 from ..utils._typedefs cimport int8_t
 
 import numpy as np
-from scipy.sparse import issparse
 
 
-cdef float64_t INFINITY = np.inf
+# Introduce a fused-class to make it possible to share the split implementation
+# between the dense and sparse cases in the node_split_best and node_split_random
+# functions. The alternative would have been to use inheritance-based polymorphism
+# but it would have resulted in a ~10% overall tree fitting performance
+# degradation caused by the overhead frequent virtual method lookups.
+ctypedef fused Partitioner:
+    DensePartitioner
```

Comment:
Since this is still a fused type, I think there should not be any v-table look ups.

But to be sure, can you check the C++ code in `node_split_best` to see if there is a v-table look up?

Reply:
The generated C++ code shows:

```
static CYTHON_INLINE int __pyx_fuse_0__pyx_f_7sklearn_4tree_9_splitter_node_split_best(struct __pyx_obj_7sklearn_4tree_9_splitter_Splitter *, struct __pyx_obj_7sklearn_4tree_12_partitioner_DensePartitioner *, struct __pyx_obj_7sklearn_4tree_10_criterion_Criterion *, struct __pyx_t_7sklearn_4tree_9_splitter_SplitRecord *, struct __pyx_t_7sklearn_4tree_5_tree_ParentInfo *); /*proto*/
static CYTHON_INLINE int __pyx_fuse_1__pyx_f_7sklearn_4tree_9_splitter_node_split_best(struct __pyx_obj_7sklearn_4tree_9_splitter_Splitter *, struct __pyx_obj_7sklearn_4tree_12_partitioner_SparsePartitioner *, struct __pyx_obj_7sklearn_4tree_10_criterion_Criterion *, struct __pyx_t_7sklearn_4tree_9_splitter_SplitRecord *, struct __pyx_t_7sklearn_4tree_5_tree_ParentInfo *); /*proto*/
static CYTHON_INLINE int __pyx_fuse_0__pyx_f_7sklearn_4tree_9_splitter_node_split_random(struct __pyx_obj_7sklearn_4tree_9_splitter_Splitter *, struct __pyx_obj_7sklearn_4tree_12_partitioner_DensePartitioner *, struct __pyx_obj_7sklearn_4tree_10_criterion_Criterion *, struct __pyx_t_7sklearn_4tree_9_splitter_SplitRecord *, struct __pyx_t_7sklearn_4tree_5_tree_ParentInfo *); /*proto*/
static CYTHON_INLINE int __pyx_fuse_1__pyx_f_7sklearn_4tree_9_splitter_node_split_random(struct __pyx_obj_7sklearn_4tree_9_splitter_Splitter *, struct __pyx_obj_7sklearn_4tree_12_partitioner_SparsePartitioner *, struct __pyx_obj_7sklearn_4tree_10_criterion_Criterion *, struct __pyx_t_7sklearn_4tree_9_splitter_SplitRecord *, struct __pyx_t_7sklearn_4tree_5_tree_ParentInfo *); /*proto*/
```

so there is no v-table lookup as there is a specialized function generated for each combination of `node_split_best/node_split_random` and `Dense/Sparse Partitioner`.

Reply:
For now, I reverted the BasePartitioner in [e24cda9](https://github.com/scikit-learn/scikit-learn/pull/29458/commits/e24cda96ae85f5200e6cd85e546e23ee98a9427a) and implemented your suggestion: https://github.com/scikit-learn/scikit-learn/pull/29458#discussion_r1684847083